### PR TITLE
use yarn.lock when building the apps

### LINKED ISSUE #3034: Verify integrity of dependencies in production with yarn
After https://github.com/mozilla/addons-frontend/issues/1391 we will be using `yarn` to install dependencies in production. 

Let's make sure all packages get installed without any surprises.

* A production install should never upgrade a package or its dependencies, even if a sub-dependency is declared with a loose version range. Please see [this detailed example](https://github.com/mozilla/addons-frontend/pull/5897#issuecomment-419574560) of what we should prevent.
* Verify the integrity of all packages. If one gets tampered with, the install step should fail. (Does yarn already do this by default? Maybe we need [`yarn check`](https://yarnpkg.com/en/docs/cli/check))

This issue is about **preventing malware on NPM from creeping into a production build** (see the [eslint malware scenario](https://www.theregister.co.uk/2018/07/12/npm_eslint/) as an example). It would still be possible for a developer to unknowingly introduce malware when upgrading a dependency -- that problem is out of scope.

### LINKED ISSUE #6194: Make deploy builds more predictable with yarn.lock
Related to #3034 

Currently, we do not use the `yarn.lock` file to download/install the dependencies in the `Dockerfile`, which means the build process does not rely on the `yarn.lock` file at all.

We should fix this by using the `yarn.lock` file when running `yarn install`.

For QA: both AMO and Disco Pane should be deployed without any problem (it will be either fine or completely broken).

### PR DESCRIPTION
**Pull request by willdurand**
Fix #6194

### CODE REVIEW

**Comment by willdurand** [comment]
I cannot make the `--offline` approach work, so I guess this is a good start.

**Comment by willdurand** [comment]
> Let me know if I'm missing something but I don't think this solves #3034 (I just updated its description btw). Aside from not solving the issue, I don't see any benefit to adding --pure-lockfile.
> 
> As far as I can tell, --pure-lockfile only stops yarn install from generating a lock file, that's it.

It is not clear to me what this option does actually. I re-read the GH issues and it now seems pointless: I don't understand why people keep mentioning it when it comes to looking for "how to install dependencies according to the lock file?". Anyway...

> I cloned the repo in a temp directory, edited the version of a dependency in package.json, and ran yarn install --pure-lockfile. It still tried to install the new dependency version from package.json.

That sounds like the correct behavior to me given that you change the `package.json` file without changing the `yarn.lock` file. More info here: https://github.com/yarnpkg/yarn/issues/5270#issuecomment-362016699. Those STR are unfortunately not matching what we want to achieve in #3034 IIRC.

Both files should be synchronized, and that's why we keep pushing `Sync yarn.lock with master` commits time to time. `yarn install` does whatever needs to be done to make sure `package.json` is satisfied, which should be the case because we synchronize both files all the time.
 
> Maybe we need --frozen-lockfile? I think that will fail if package.json is introducing new modules/versions. Getting a failure like this in production/dev/stage is too late, though.

Adding such an option on Travis means we can stop using Greenkeeper because all PRs will be red :/

We could potentially add a Travis job for builds on `master`, but that means we cannot merge Greenkeeper PRs directly from the web UI (which is fine to me, though).

> Anyway, I'm not convinced that either option will solve this part of #3034 :
> 
>     A production install should never upgrade a package (or its dependencies) beyond what's in our lock file.

I tend to disagree: that will be the case with this PR given that it adds the `yarn.lock` file when installing the dependencies in the Docker image build process.

> I think we need to consider a new approach. We either need an offline mirror or something like shrinkwrap.

Offline mode is probably too much honestly, we would have to commit 50-60MB of `tar` files and maintain them. This is probably the best paranoid option, though.

We currently don't use the `yarn.lock` file at all, we are very lucky that everything works. Having that should partially fix the issue mentioned in the OP.

The second part could be fixed by running `yarn check` during the Docker image build, which is used by the UI tests IIRC so we would see the errors in travis-ci.

WDYT?

**Comment by willdurand** [comment]
@kumar303 ^

**Comment by bobsilverberg** [comment]
@willdurand Do we need to keep this PR open, or are we looking at other alternatives to solve the issue?

**Comment by willdurand** [comment]
> @willdurand Do we need to keep this PR open, or are we looking at other alternatives to solve the issue?

I think it should stay open until I get the chance to talk to @kumar303 about it again, hopefully I'll have an update by the end of this week.

**Comment by kumar303** [comment]
Sorry for missing the last comment.

I'll come back to this requirement again:

> A production install should never upgrade a package or its dependencies beyond what's in our lock file (or something else that declares all package versions).

You mentioned in the last comment that your patch will fix this. How?

Let's take `isomorphic-fetch` as an example. We require a specific version of it (not a loose version):

https://github.com/mozilla/addons-frontend/blob/9f14a306602179c0ac4f38b1335a77b234b4ef0d/package.json#L184

But in the lock file, its dependency on `node-fetch` is loose (if I understand how lock files work--maybe I don't!):

https://github.com/mozilla/addons-frontend/blob/9f14a306602179c0ac4f38b1335a77b234b4ef0d/yarn.lock#L4992-L4997

What if an attacker hacks an account and publishes `1.0.2` of `node-fetch` with a coin miner in it? Won't the next production install of `addons-frontend` pull in `1.0.2` of `node-fetch`? If not, how does your patch prevent that?

**Comment by kumar303** [comment]
> A production install should never upgrade a package or its dependencies beyond what's in our lock file (or something else that declares all package versions).

I also just realized that I worded this incorrectly. What I mean is to never upgrade any *sub-dependency* beyond its version that existed at the time we pinned the top-level *dependency*.

We accomplish this in Python by pinning *all* package versions (even sub-dependencies) and installing the requirements with [`--no-deps`](https://github.com/mozilla/addons-server/blob/f922c5e1f212f71eceac8d3fd5a620d1b8a17032/Dockerfile.deploy#L73).

**Comment by willdurand** [comment]
> You mentioned in the last comment that your patch will fix this. How?

@kumar303 mainly by using the `yarn.lock` file, this file is not used yet so it is a miracle that things work and get installed properly. It can only be better. Yarn folks say that the `yarn.lock` file ensures reproducible builds (and yarn in general, but does that mean it is SemVer-compliant? Or better?)

According to [this issue](https://github.com/yarnpkg/yarn/issues/4814), `frozen-lockfile` will only fail if `package.json` and `yarn.lock` files are not in sync, which is not what we want either.

> I also just realized that I worded this incorrectly. What I mean is to never upgrade any sub-dependency beyond its version that existed at the time we pinned the top-level dependency.
> 
> We accomplish this in Python by pinning all package versions (even sub-dependencies) and installing the requirements with --no-deps.

Offline mirror is probably the feature that allows that: https://yarnpkg.com/blog/2016/11/24/offline-mirror/.

To sum up, I really think we need that `yarn.lock` file as part of the build process, that will be better than without. We can still open a separate issue for the paranoid mode (a.k.a. pinning all package versions of everything). WDYT?

**Comment by kumar303** [comment]
> I really think we need that yarn.lock file as part of the build process, that will be better than without. 

I agree. If you want to merge your patch as is then that's fine with me. I don't think it should close https://github.com/mozilla/addons-frontend/issues/3034 but you could make another issue to track it like "Make deploy builds more predictable with `yarn.lock`" or something.

> We can still open a separate issue for the paranoid mode (a.k.a. pinning all package versions of everything).

I intended https://github.com/mozilla/addons-frontend/issues/3034 to be that issue and I think the bullet point about making sure no packages can be upgraded captures that intent. I will update the issue to clarify what I meant in that sentence.

**Comment by willdurand** [comment]
@kumar303 alrighty, updated PR title/desc. and created a different issue. r?
